nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -3084,14 +3084,85 @@ impl ArrayPrototype {
         call_function(agent, gc, func, array.into_value(), None)
     }
 
+    /// ### [23.1.3.37 Array.prototype.unshift ( ...items )](https://tc39.es/ecma262/#sec-array.prototype.unshift)
+    ///
+    /// This method prepends the arguments to the start of the array, such that
+    /// their order within the array is the same as the order in which they appear
+    /// in the argument list.
+    ///
+    /// > ### Note
+    /// >
+    /// > This method is intentionally generic; it does not require that its
+    /// > this value be an Array. Therefore it can be transferred to other
+    /// > kinds of objects for use as a method.
     fn unshift(
-        _agent: &mut Agent,
-        _gc: GcScope<'_, '_>,
-
-        _this_value: Value,
-        _: ArgumentsList,
+        agent: &mut Agent,
+        mut gc: GcScope<'_, '_>,
+        this_value: Value,
+        items: ArgumentsList,
     ) -> JsResult<Value> {
-        todo!();
+        // 1. Let O be ? ToObject(this value).
+        let o = to_object(agent, this_value)?;
```

Comment:
**suggestion:** It's possible to add a fast-path here if `this_value` is an Array and it is trivial (contains no non-default descriptors, so no getters, setters, or unwritable, unconfigurable, or unenumerable properties) and dense (contains no holes), and its length property is writable. We'll first need to do a `array.reserve(agent, array.len() + items.len())`, then change the length, access the memory as a mutable slice, copy within it to move old items towards the end, and finally copy data from `items` to the beginning of the memory slice.